### PR TITLE
Optimize InstanceNormGradientOp

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary: Optimize InstanceNormGradientOp

Differential Revision: D16023630

### CODE REVIEW

**Comment by BIT-silence** [comment]
Benchmarks:

CPU with [N, C, H, W] = [128, 256, 56, 56],
NCHW order: 616ms -> 128ms
NHWC order: 1612ms -> 174ms

GPU with [N, C, H, W] = [128, 256, 112, 112],
NCHW order: 6450ms -> 37ms
NHWC order: 1419ms -> 82ms

File: caffe2/python/operator_test/instance_norm_test.py [hunk 1]
```diff
@@ -76,7 +76,16 @@ def test_instance_norm_gradients(
             store_inv_stdev=store_inv_stdev,
             epsilon=epsilon,
             order=order)
-        input_blobs = self._get_inputs(N, C, H, W, order)
+
+        input_data = np.arange(N * C * H * W).astype(np.float32)
```
Comment (houseroad) [inline caffe2/python/operator_test/instance_norm_test.py]: why moved away from `_get_inputs`?

File: caffe2/operators/instance_norm_op.h [hunk 2]
```diff
@@ -127,44 +127,148 @@ class InstanceNormOp final : public Operator<Context> {
...
+    const int ndim = X.dim();
+    const int64_t N = X.dim(0);
+    const int64_t C = order_ == StorageOrder::NCHW ? X.dim(1) : X.dim(ndim - 1);
+    const int64_t HxW = X.numel() / (N * C);
+    CAFFE_ENFORCE_EQ(gamma.numel(), C);
+    const T* dY_data = dY.template data<T>();
+    const T* X_data = X.template data<T>();
+    const T* gamma_data = gamma.template data<T>();
+    const T* mean_data = nullptr;
+    const T* rstd_data = nullptr;
```
Comment (houseroad) [inline caffe2/operators/instance_norm_op.h]: Also check the input number?

**Review by houseroad — APPROVED** [review]
Looks good.

File: caffe2/python/operator_test/instance_norm_test.py (hunk 1 above)
Comment (BIT-silence) [reply caffe2/python/operator_test/instance_norm_test.py]: _get_inputs used rand to generate input, whose range is pretty small and may lead to numerical failures for grad check. Here we use arrange and shuffle to generate a tensor those elements are far away from each other.

Comment (houseroad) [reply caffe2/python/operator_test/instance_norm_test.py]: I see, shall we just extend _get_inputs a bit? Since most of the code looks very similar

Comment (BIT-silence) [reply caffe2/python/operator_test/instance_norm_test.py]: The correct thing is just remove _get_inputs at all. There is no need for 3 lines in such a function.

File: caffe2/operators/instance_norm_op.h (hunk 2 above)
Comment (BIT-silence) [reply caffe2/operators/instance_norm_op.h]: Done.

**Pull request closed without merging**